### PR TITLE
[Fix]Local audio waveform not always working

### PR DESCRIPTION
**Pull request by ipavlidakis**
### 🔗 Issue Links

Resolves https://linear.app/stream/issue/IOS-1078/audio-waveform-isnt-always-updating

### 🎯 Goal

This revision aims to provide alignment between the AudioRecorder - that is being used to provide the local participant's waveform visualisation - and the underline RTCAudioStore.

### 🛠 Implementation

The StreamCallAudioRecorder has been updated and linked to the underline RTCAudioStore.

### 🧪 Manual Testing Notes

- Join many calls (one after another) and ensure that your audio waveform always works as expected.

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] This change should receive manual QA
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (tutorial, CMS)

### CODE REVIEW

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/CallAudioRecordingStore.swift [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            .eraseToAnyPublisher()
+    }
+
+    func dispatch(
+        _ action: CallAudioRecordingAction,
+        file: StaticString = #file,
+        function: StaticString = #function,
+        line: UInt = #line
+    ) {
+        processingQueue.addTaskOperation { [weak self] in
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/CallAudioRecordingStore.swift]: Not sure if it makes sense to have some basic store protocol with default implementations for dispatch and perform. Atm, we would need to maintain 2 implementations, potentially more in the future.

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/AVAudioRecorderBuilder.swift [hunk 2]
```diff
@@ -37,7 +37,7 @@ final class AVAudioRecorderBuilder {
     var result: AVAudioRecorder? { cachedResult }
 
     init(
-        inCacheDirectoryWithFilename filename: String,
+        inCacheDirectoryWithFilename filename: String = "recording.wav",
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/AudioSession/AudioRecorder/AVAudioRecorderBuilder.swift]: why do we set the default value?

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+ApplicationStateMiddleware.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            file: StaticString,
+            function: StaticString,
+            line: UInt
+        ) {
+            /* No-op */
+        }
+
+        // MARK: - Private Helpers
+
+        private func didUpdate(_ applicationState: ApplicationState) async {
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+ApplicationStateMiddleware.swift]: seems like this param is unused

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+CategoryMiddleware.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        init(_ store: CallAudioRecordingStore) {
+            self.store = store
+
+            cancellable = audioStore
+                .publisher(\.category)
+                .filter { $0 != .playAndRecord || $0 != .record }
+                .sink { [weak self] _ in self?.store?.dispatch(.setIsRecording(false)) }
+        }
+
+        func apply(
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+CategoryMiddleware.swift]: do we need this method? I mostly see no-op implementations.

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/CallAudioRecordingStore.swift [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            if case let .setMeter(value) = action {
+                metersUpdated.append(value)
+                guard
+                    metersUpdated.endIndex == metersUpdatedLimit
+                else {
+                    return
+                }
+
+                let sum = metersUpdated.reduce(0, +)
+                let average = sum / Float(metersUpdatedLimit)
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/CallAudioRecordingStore.swift]: how frequently do we receive updates? computing average on 500 is translated into how long intervals?

**Review by martinmitrevski — COMMENTED** [review]
Looks good, left few comments. Let's test this on the calls.

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/AVAudioRecorderBuilder.swift (hunk 2 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Utils/AudioSession/AudioRecorder/AVAudioRecorderBuilder.swift]: Before we were also setting the value but on the `InjectValue.currentValue`. Even though we can initialize and AVAudioRecorder without a file, it seems that this audioRecorder doesn't produce any meters.

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+ApplicationStateMiddleware.swift (hunk 3 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+ApplicationStateMiddleware.swift]: It isn't but as the triggering event is the applicationState change, i have added it for reference. Do you prefer to remove it?

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+CategoryMiddleware.swift (hunk 4 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+CategoryMiddleware.swift]: Yes, the method is necessary in the Redux scope. The reason we aren't using it in a few of the current middleware is because those instances are being used to synchronise values from outside the store, with the store.

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/CallAudioRecordingStore.swift (hunk 5 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/CallAudioRecordingStore.swift]: As we discussed, this is only being used for logging purposes and to reduce the noise in our logs. Currently it's happening 1-2 seconds

File: Sources/StreamVideo/Utils/Store/StoreNamespace.swift [hunk 6]
```diff
@@ -0,0 +1,46 @@
+//
+// Copyright © 2025 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+protocol StoreNamespace {
```
Comment (ipavlidakis) [inline Sources/StreamVideo/Utils/Store/StoreNamespace.swift:42]: This setup will allow us not only to instantiate stores easily (and configure them) but also to mock them for unit tests (as we can replace the executor with a new implementation and spy every execution).

File: Sources/StreamVideo/Utils/Store/Store.swift [hunk 7]
```diff
@@ -0,0 +1,147 @@
+//
+// Copyright © 2025 Stream.io Inc. All rights reserved.
+//
+
+import Combine
+import Foundation
+
+final class Store<Namespace: StoreNamespace>: @unchecked Sendable {
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/Store/Store.swift:46]: Looks very nice!

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+ApplicationStateMiddleware.swift (hunk 3 above)
Comment (martinmitrevski) [reply Sources/StreamVideo/Utils/AudioSession/AudioRecorder/CallAudioRecordingStore/Middleware/CallAudioRecordingStore+ApplicationStateMiddleware.swift]: might be confusing yes - maybe just rename the method to `didUpdateApplicationState`?

File: StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    }
+    
+    // MARK: - State Read/Write Race Tests
+    
+    /// Tests reading state while it's being updated.
+    func test_stateReadDuringWrite_isConsistent() async throws {
+        // Given
+        let iterations = 1000
+        nonisolated(unsafe) var capturedStates: [RaceTestState] = []
+        let lock = UnfairQueue()
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift:41]: The usage of propertyWrappers inside a function body feels weird tbh. That's why i'm using it. The end result should be the same though.

File: Sources/StreamVideo/Utils/AudioSession/AudioRecorder/Extensions/AVAudioRecorder+Convenience.swift [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            AVSampleRateKey: 12000,
+            AVNumberOfChannelsKey: 1,
+            AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue
+        ]
+    ) throws -> AVAudioRecorder {
+        // Get the cache directory for temporary audio storage
+        let documentPath = FileManager.default.urls(
+            for: .cachesDirectory,
+            in: .userDomainMask
+        )[0]
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/AudioSession/AudioRecorder/Extensions/AVAudioRecorder+Convenience.swift]: isn't it better to use .first here and throw an error if it's nil? Theoretically, this could crash.

File: Sources/StreamVideo/Utils/Store/Middleware.swift [hunk 10]
```diff
@@ -0,0 +1,89 @@
+//
+// Copyright © 2025 Stream.io Inc. All rights reserved.
+//
+
+/// Base class for store middleware that intercept actions for side effects.
+///
+/// Middleware sit between action dispatch and state reduction, allowing you
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/Store/Middleware.swift]: *sits

File: Sources/StreamVideo/Utils/Store/Middleware.swift [hunk 11]
```diff
@@ -0,0 +1,89 @@
+//
+// Copyright © 2025 Stream.io Inc. All rights reserved.
+//
+
+/// Base class for store middleware that intercept actions for side effects.
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/Store/Middleware.swift]: *intercepts

File: Sources/StreamVideo/Utils/Store/Middleware.swift [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+/// to:
+/// - Perform asynchronous operations
+/// - Dispatch additional actions
+/// - Access external services
+/// - Transform or filter actions
+/// - Log or monitor state changes
+///
+/// ## Architecture
+///
+/// Middleware receive actions before reducers process them. They can:
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/Store/Middleware.swift]: *receives

File: Sources/StreamVideo/Utils/Store/Store.swift [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+        initialState: Namespace.State,
+        reducers: [Reducer<Namespace>],
+        middleware: [Middleware<Namespace>],
+        logger: StoreLogger<Namespace>,
+        executor: StoreExecutor<Namespace>
+    ) {
+        self.identifier = identifier
+        stateSubject = .init(initialState)
+        self.reducers = reducers
+        self.middleware = []
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/Store/Store.swift:95]: why is this set to empty array?

File: Sources/StreamVideo/Utils/Store/StoreLogger.swift [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+/// A logger for recording store operations and debugging state changes.
+///
+/// The store logger provides visibility into the action processing
+/// pipeline, logging successful completions and failures. It can be
+/// subclassed to customize logging behavior for specific stores.
+///
+/// ## Customization
+///
+/// Subclass this logger to:
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Utils/Store/StoreLogger.swift:15]: Seems like a useful thing to expose to customers, wdyt?

File: StreamVideoTests/Utils/Store/Store_ConcurrencyTests.swift [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+import XCTest
+
+/// Comprehensive tests for Store thread safety and concurrency.
+///
+/// These tests verify that the Store correctly handles:
+/// - Concurrent action dispatches
+/// - State consistency under high load
+/// - Order preservation for serial processing
+/// - Publisher behavior under concurrent access
+final class Store_ConcurrencyTests: XCTestCase, @unchecked Sendable {
```
Comment (martinmitrevski) [inline StreamVideoTests/Utils/Store/Store_ConcurrencyTests.swift:17]: Great stuff! 👏

File: StreamVideoTests/Utils/Store/Store_PerformanceTests.swift [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+            
+            // Reset
+            store.dispatch(.reset)
+        }
+    }
+    
+    // MARK: - Complex State Tests
+    
+    /// Measures performance with complex state updates.
+    func test_measureComplexStateUpdates() {
```
Comment (martinmitrevski) [inline StreamVideoTests/Utils/Store/Store_PerformanceTests.swift:155]: looks great! Just one question - do you think these might become flaky depending on which machine they're run on?

File: StreamVideoTests/Utils/Store/Store_PerformanceTests.swift [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+        }
+        
+        // Clean up
+        middleware.forEach { store.remove($0) }
+    }
+    
+    // MARK: - Memory Performance Tests
+    
+    /// Tests memory usage with large state.
+    func test_memoryUsageWithLargeState() {
```
Comment (martinmitrevski) [inline StreamVideoTests/Utils/Store/Store_PerformanceTests.swift:226]: another clever one!

File: StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+import XCTest
+
+/// Tests for specific race conditions and edge cases in Store.
+///
+/// These tests target specific threading scenarios that could cause issues:
+/// - Read/write races
+/// - Middleware state access during updates
+/// - Publisher timing issues
+/// - Cleanup during active operations
+final class Store_RaceConditionTests: XCTestCase, @unchecked Sendable {
```
Comment (martinmitrevski) [inline StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift:17]: Store testing final boss

File: StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinmitrevski) [reply StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift:41]: nah, don't listen to it 😅

File: StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+    override func tearDown() {
+        cancellables = nil
+        store = nil
+        super.tearDown()
+    }
+    
+    // MARK: - State Read/Write Race Tests
+    
+    /// Tests reading state while it's being updated.
+    func test_stateReadDuringWrite_isConsistent() async throws {
```
Comment (martinmitrevski) [inline StreamVideoTests/Utils/Store/Store_RaceConditionTests.swift:37]: Tests like these are really useful. Thinking if we should generalize them at some point, to test race conditions elsewhere (other parts of the codebase, other products). Maybe move them to core testing utils. 
Anyway, not in scope for this PR.

File: Sources/StreamVideo/Utils/Store/Middleware.swift (hunk 11 above)
Comment (martinmitrevski) [reply Sources/StreamVideo/Utils/Store/Middleware.swift]: Few typos here, which I think you do on purpose, since you consider middleware as plural, but in general it should be singular.
Anyway, not important for sure, feel free to ignore these.

**Review by martinmitrevski — APPROVED** [review]
Really nice and elegant implementation! Left few small comments, but there's no need to re-review.
Let's test this a bit more before merging.

File: StreamVideoTests/Utils/Store/Store_PerformanceTests.swift (hunk 16 above)
Comment (ipavlidakis) [reply StreamVideoTests/Utils/Store/Store_PerformanceTests.swift:155]: Given that on CI we use m machines i think we are safe. But even if it starts happening we can adapt the acceptable threshold.

File: Sources/StreamVideo/Utils/Store/StoreLogger.swift (hunk 14 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Utils/Store/StoreLogger.swift:15]: You mean the logger or the whole store implementation?

File: Sources/StreamVideo/Utils/Store/Store.swift (hunk 13 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Utils/Store/Store.swift:95]: Middleware requires installation when they are added to the store (the dispatcher and stateProvider). This logic exists in the addMiddleware method and i didn't want to duplicate it. Let me add a comment to clarify the intent.
